### PR TITLE
Add unit tests for Engine to Python format conversion

### PR DESCRIPTION
**Pull request by par4m**
Implements https://github.com/cocoindex-io/cocoindex/issues/185

### CODE REVIEW

File: python/cocoindex/tests/test_convert.py [hunk 1]
```diff
@@ -67,3 +80,81 @@ def test_to_engine_value_tuple():
...
+        (bool, True),
+        # (type(None), None),  # Removed unsupported NoneType
+    ]:
+        engine_type = encode_enriched_type(py_type)["type"]
+        converter = make_engine_value_converter([], engine_type, py_type)
+        assert converter(value) == value
+
+def test_make_engine_value_converter_struct():
+    engine_type = encode_enriched_type(Order)["type"]
+    converter = make_engine_value_converter([], engine_type, Order)
```
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: For all these tests, we're calling `encode_enriched_type()` and `make_engine_value_converter()` in sequence. It'll be great if we wrap these two to a function and call this to build converter for test. It will make each test slightly easier to read.

File: python/cocoindex/tests/test_convert.py [hunk 2]
```diff
@@ -19,6 +25,21 @@ class Basket:
...
     order: Order
+    tags: list[Tag] = None
+
+@dataclass
+class NestedStruct:
+    customer: Customer
+    orders: list[Order]
+    count: int = 0
+
+def build_converter(py_type, target_type=None):
```
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: The first argument `py_type` is actually used to simulate type on the engine side, so something like `engine_type_in_py` maybe more clear.

File: python/cocoindex/tests/test_convert.py [hunk 3]
```diff
@@ -67,3 +88,74 @@ def test_to_engine_value_tuple():
...
+        (float, 3.14),
+        (str, "hello"),
+        (bool, True),
+        # (type(None), None),  # Removed unsupported NoneType
+    ]:
+        converter = build_converter(py_type)
+        assert converter(value) == value
+
+def test_make_engine_value_converter_struct():
+    converter = build_converter(Order)
```
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: For several cases, like extra fields, missing fields, order difference, the actual reason to trigger these case is: engine side has a slightly different schema from Python type definitions (the schema from engine is determined based on source/function output types + flow definitions).

To simulate such situations in this test, we may need to define separate python types and pass two different types to `build_converter` (one to simulate type on engine side, one is for type on Python side, e.g. to cover the case of different field orders, they can be two different dataclasses with different field orders).

File: python/cocoindex/tests/test_convert.py (hunk 2 above)
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: The name `build_engine_value_converter` may be more clear.

File: python/cocoindex/tests/test_convert.py [hunk 4]
```diff
@@ -153,9 +154,66 @@ def test_make_engine_value_converter_collections():
...
+    """
+    @dataclass
+    class EngineOrder:
+        id: str
+        name: str
+    @dataclass
+    class PythonOrder:
+        id: str
+        name: str
+        price: float = 0.0
```
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: The extra field and missing field cases are not limited to the fields at last. If we can make the extra / missing field in the middle, the test will catch more potential problems.

Comment (par4m) [reply python/cocoindex/tests/test_convert.py]: There's a decorator in pytest which can be used for this :- 
```python

@pytest.mark.parametrize(
    "engine_fields, python_fields, python_defaults, engine_val, expected_python_val",
    [
        # Extra field - middle
        (
            [("id", str), ("name", str)],
            [("id", str), ("price", float), ("name", str)],
            {"price": 0.0},
            ["O123", "mixed nuts"],
            ("O123", 0.0, "mixed nuts"),
        ),
        # Missing field - middle
        (
            [("id", str), ("price", float), ("name", str)],
            [("id", str), ("name", str)],
            {},
            ["O123", 25.0, "mixed nuts"],
            ("O123", "mixed nuts"),
        ),
        # Extra field - start
        (
            [("name", str), ("price", float)],
            [("extra", str), ("name", str), ("price", float)],
            {"extra": "default"},
            ["mixed nuts", 25.0],
            ("default", "mixed nuts", 25.0),
        ),
        # Missing field - start
        (
            [("extra", str), ("name", str), ("price", float)],
            [("name", str), ("price", float)],
            {},
            ["unexpected", "mixed nuts", 25.0],
            ("mixed nuts", 25.0),
        ),
    ]
)
```
Does this cover all the edge cases ?

Comment (badmonster0) [reply python/cocoindex/tests/test_convert.py]: Yes! This decorator sounds great!

File: python/cocoindex/tests/test_convert.py [hunk 5]
```diff
@@ -67,3 +89,193 @@ def test_to_engine_value_tuple():
...
+def test_make_engine_value_converter_defaults_and_missing_fields():
+    # Missing optional field in engine value
+    converter = build_engine_value_converter(Customer)
+    engine_val = ["Alice", ["O1", "item1", 10.0, "default_extra"], None]  # tags explicitly None
+    assert converter(engine_val) == Customer("Alice", Order("O1", "item1", 10.0, "default_extra"), None)
+    # Extra field in engine value (should ignore)
+    engine_val = ["Alice", ["O1", "item1", 10.0, "default_extra"], [["vip"]], "extra"]
+    assert converter(engine_val) == Customer("Alice", Order("O1", "item1", 10.0, "default_extra"), [Tag("vip")])
+
+def test_engine_python_schema_field_order():
```
Comment (badmonster0) [inline python/cocoindex/tests/test_convert.py]: Since we have the generic parameterized `test_field_position_cases`, seems this case can be folded into that one as another parameter set? (the same for `test_engine_python_schema_extra_field` and `test_engine_python_schema_missing_field`) This will reduce code for these tests and make them more easily understandable and maintainable.

Comment (par4m) [reply python/cocoindex/tests/test_convert.py]: Following this, scenarios from `test_make_engine_value_converter_struct` and `test_make_engine_value_converter_defaults_and_missing_fields` are also parameterized in a single test: `test_struct_conversion_cases`. Is there anything else that needs refactoring ?

**Comment by badmonster0** [comment]
Thanks a lot for making this test!
